### ISSUE TITLE
fix_options should run after chunk hooks

### ISSUE THREAD

**Issue by atusy** [yihui/knitr#1876]
I set chunk options and hooks so that I can specify `fig.width` and `fig.height` in centimeters.
What I did is to divide `fig.width/height` by 2.54.
In order to keep the default size of figures, I also set chunk options to multiply default `fig.width/height` by 2.54.
However, I realized that `out.width` is calculated before dividing `fig.width/height` by 2.54 in `rmarkdown::html_document`.

It happens when `keep_md` is `FALSE`.
I read the source code and found that `fix_options` is calculating `out.width` before applying chunk hooks on `fig.width`.
I suggest run `fix_options` after chunk hooks, though I'm not sure about side effects...

When `keep_md` is TRUE, `fig.retina` becomes `NULL` and avoids the problem.
https://github.com/rstudio/rmarkdown/blob/ff1b279e795b62b1ffeabcc9aa5bf7386a7ebb83/R/html_document.R#L475-L476

I will draft a PR and see if it passes the test.

# Reprex

```r
rmarkdown::render('example.Rmd')
```

<details><summary>example.Rmd</summary>

````md
---
title: "Untitled"
output: html_document
---

```{r}
f = knitr::knit_hooks$get('plot')
knitr::knit_hooks$set(
  plot = function(x, options) {
    print(options$out.width)
    f(x, options)
  }
)
```


```{r before-hook}
knitr::opts_current$get()[c('fig.height', 'fig.width')]
plot(iris)
```

```{r hook}
knitr::opts_chunk$set(
  fig.height = knitr::opts_chunk$get('fig.height') * 2.54,
  fig.width = knitr::opts_chunk$get('fig.width') * 2.54
)

knitr::opts_hooks$set(
  fig.height = function(options) {
    options$fig.height <- options$fig.height / 2.54
    options
  },
  fig.width = function(options) {
    options$fig.width <- options$fig.width / 2.54
    options
  }
)
```



```{r after-hook}
knitr::opts_current$get()[c('fig.height', 'fig.width')]
plot(iris)
```
````

</details>

<details><summary>
Result: out.width differs before (672) and after (1706.88) setting chunk hooks
</summary>

````
rmarkdown::render('example.Rmd')
Loading knitr


processing file: example.Rmd
  |..........                                                                                    |  11%
  ordinary text without R code

  |.....................                                                                         |  22%
label: unnamed-chunk-1
  |...............................                                                               |  33%
  ordinary text without R code

  |..........................................                                                    |  44%
label: before-hook
[1] 672
  |....................................................                                          |  56%
  ordinary text without R code

  |...............................................................                               |  67%
label: hook
  |.........................................................................                     |  78%
  ordinary text without R code

  |....................................................................................          |  89%
label: after-hook
[1] 1706.88
  |..............................................................................................| 100%
  ordinary text without R code


output file: example.knit.md

/usr/bin/pandoc +RTS -K512m -RTS example.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output example.html --email-obfuscation none --self-contained --standalone --section-divs --template /usr/local/lib/R/site-library/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable 'theme:bootstrap' --include-in-header /tmp/Rtmp1l8BoQ/rmarkdown-strc05a73c9921a.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --lua-filter /usr/local/lib/R/site-library/rmarkdown/rmd/lua/pagebreak.lua --lua-filter /usr/local/lib/R/site-library/rmarkdown/rmd/lua/latex-div.lua 

Output created: example.html
````

</details>

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by atusy** [comment]
There is another reason I consider `fix_options` should run after hooks.
As far as I understand, `fix_options` verifies and corrects the chunk options.
By running `fix_options` after hooks, we can correct malformed chunk options generated by the hooks.

**Comment by yihui** [comment]
Resolved by #1877. Thanks!
